features/cellular/framework/common/APN_db.cpp

```diff
@@ -0,0 +1,182 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ublox, ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* ----------------------------------------------------------------
+   APN stands for Access Point Name, a setting on your modem or phone
+   that identifies an external network your phone can access for data
+   (e.g. 3G or 4G Internet service on your phone).
+
+   The APN settings can be forced when calling the join function.
+   Below is a list of known APNs that us used if no apn config
+   is forced. This list could be extended by other settings.
+
+   For further reading:
+   wiki apn: http://en.wikipedia.org/wiki/Access_Point_Name
+   wiki mcc/mnc: http://en.wikipedia.org/wiki/Mobile_country_code
+   google: https://www.google.de/search?q=APN+list
+---------------------------------------------------------------- */
+
+#include "APN_db.h"
+
+#if (MBED_CONF_CELLULAR_USE_APN_LOOKUP || (NSAPI_PPP_AVAILABLE && MBED_CONF_PPP_CELL_IFACE_APN_LOOKUP))
+
+/**
+ * Helper to generate the APN string
+ */
+#define _APN(apn,username,password) apn "\0" username "\0" password "\0"
+
+/**
+ * APN lookup struct
+ */
+typedef struct {
+    const char *mccmnc; /**< mobile country code (MCC) and mobile network code MNC */
+    const char *cfg;     /**<  APN configuartion string, use _APN macro to generate */
+} APN_t;
+
+/**
+ * Default APN settings used by many networks
+ */
+static const char *apndef = _APN("internet",,);
```

Comment:
Can save a little more here by making that a `char[]` array like apnlut. It's currently a RAM pointer to a ROM char array.

Reply:
Do you mean `static const char apndef[] = _APN("internet",,);` ? At least with GCC this does not change anything.

Reply:
Yes.

I guess that as `apndef` is a file-scope variable, the compiler manages to figure out that no-one ever writes to it, so it can optimise it out, and treat it as a constant, even though it's not declared as such.

I'd still prefer to say what we actually need, rather than rely on the compiler being bright. If you don't need a pointer, don't declare one. Assuming non-automatic storage,

    const char xx[] = "foo"; // 4 bytes of string in ROM

is better than

    const char * const xx = "foo"; // 4 bytes of string in ROM, and a 4-byte pointer to it in ROM

is better than

    const char *xx = "foo"; // 4 bytes of string in ROM, 4 bytes of pointer initialisation in ROM, 4 bytes of pointer in RAM

Reply:
Fixed